### ISSUE TITLE
newDeref: possible false-positive with generics

### ISSUE THREAD

**Issue by junk1tm** [go-critic/go-critic#1272]
First of all, thank you for the awesome linter, I'm in love with go-critic ❤️

It seems like [newDeref](https://go-critic.com/overview.html#newderef) reports a false-positive when type parameters are used:

```go
func test[T any]() T {
	return *new(T)
}
```

> newDeref: replace `*new(T)` with `T(nil)`

Which is not possible because the conversion is invalid:

> cannot convert nil to T ​compiler: InvalidConversion

Found [here](https://github.com/junk1tm/assert/blob/main/assert.go#L105).

golangci-lint version: 1.49.0
gocritic version: v0.6.5 (installed with brew)

**Comment by junk1tm** [comment]
@quasilyte I'd like to work on this one if you don't mind

Would it be ok for `newDeref` to completely ignore type parameters?

**Comment by quasilyte** [comment]
Sure, go ahead. It's always nice to have contributions. :)

> Would it be ok for newDeref to completely ignore type parameters?

I would say yes. I can't figure out a better option right now.

@cristaloleg do you have any ideas about it?

**Comment by cristaloleg** [comment]
Yeah, sounds reasonable.

**Comment by quasilyte** [comment]
@junk1tm it's more problematic since the replacement is no longer an expression.
It also introduces a new variable.

I wonder if Go compiler optimizes `*new(T)` expressions and avoids redundant `newobject` calls. :)
If it doesn't, maybe it's time for it to learn such idiom that may become more useful with generics.

**Comment by junk1tm** [comment]
Seems like the following code when compiled via `go tool compile -S main.go` does not produce any `runtime.newobject` calls 🤔 

```go
func main() {
	_ = test[time.Time]()
}

//go:noinline
func test[T fmt.Stringer]() string {
	s := (*new(T)).String()
	return s
}

```

@quasilyte does it mean that the optimization you mentioned actually exists? (I'm a compiler newbie 😅)

`go tool compile -m main.go` reports that `main.go:14:12: new() does not escape`
